File: pi-repomap/ts-src/engine/rank.ts
import type { Symbol } from "../models.js";
import { isTestFile } from "./discover.js";
import { readText } from "./utils.js";

const IDENTIFIER_RE = /[a-zA-Z_]\w+/g;

export function buildSymbolIndex(allSymbols: Record<string, Symbol[]>): Record<string, Symbol[]> {
  const index = new Map<string, Symbol[]>();
  for (const symbols of Object.values(allSymbols)) {
    for (const symbol of symbols) {
      for (const token of referenceTokens(symbol.name)) {
        const entries = index.get(token) ?? [];
        entries.push(symbol);
        index.set(token, entries);
      }
    }
  }
  return Object.fromEntries(index.entries());
}

function referenceTokens(name: string): Set<string> {
  const base = name.includes(".") ? name.split(".").at(-1) ?? name : name;
  const tokens = new Set<string>([base]);
  if (!name.includes(".")) {
    tokens.add(name);
  }
  return new Set([...tokens].filter((token) => /^[a-zA-Z_]\w+$/.test(token)));
}

export async function computeImportance(
  allSymbols: Record<string, Symbol[]>,
  repoPath: string,
  fileInrefs: Record<string, number> = {},
): Promise<void> {
  const index = buildSymbolIndex(allSymbols);
  const tokenSet = new Set(Object.keys(index));
  if (tokenSet.size === 0) {
    return;
  }

  const refCount = new Map<string, number>();
  for (const filePath of Object.keys(allSymbols)) {
    const content = await readText(repoPath, filePath, Number.MAX_SAFE_INTEGER);
    if (!content) {
      continue;
    }

    const wordsInFile = new Set<string>();
    for (const match of content.matchAll(IDENTIFIER_RE)) {
      const token = match[0];
      if (tokenSet.has(token) && !wordsInFile.has(token)) {
        wordsInFile.add(token);
        for (const symbol of index[token] ?? []) {
          if (symbol.file !== filePath) {
            const key = `${symbol.file}::${symbol.name}`;
            refCount.set(key, (refCount.get(key) ?? 0) + 1);
          }
        }
      }
    }
  }

  for (const [filePath, symbols] of Object.entries(allSymbols)) {
    const isTest = isTestFile(filePath);
    for (const symbol of symbols) {
      const key = `${symbol.file}::${symbol.name}`;
      let score = refCount.get(key) ?? 0;
      symbol.refCount = score;

      if (symbol.kind === "class" || symbol.kind === "interface") {
        score *= 1.5;
      } else if (symbol.kind === "resource" || symbol.kind === "module" || symbol.kind === "data") {
        score *= 2.0;
      } else if (symbol.kind === "key") {
        score = -1.0;
      }

      const baseName = symbol.name.includes(".") ? symbol.name.split(".").at(-1) ?? symbol.name : symbol.name;
      if (["main", "index", "App", "Server", "setup", "configure", "create_app", "handler"].includes(baseName)) {
        score += 5.0;
      }

      score += Math.min(fileInrefs[filePath] ?? 0, 25) * 0.35;
      if (isTest || baseName.startsWith("test_") || baseName.startsWith("it(")) {
        score *= 0.05;
      }

      symbol.importance = score;
    }
  }
}

export function suggestedReads(allSymbols: Record<string, Symbol[]>, files: string[], limit = 5): string[] {
  const scored = Object.entries(allSymbols)
    .map(([filePath, symbols]) => ({
      filePath,
      total: symbols.reduce((sum, symbol) => sum + symbol.importance, 0),
    }))
    .sort((left, right) => right.total - left.total || left.filePath.localeCompare(right.filePath));

  if (scored.length === 0) {
    return files.slice(0, limit);
  }

  return scored.slice(0, limit).map((entry) => entry.filePath);
}

File: pi-repomap/src/scope/engine/rank.py
from __future__ import annotations

import os
import re
from collections import defaultdict
from typing import Dict, List, Optional, Set, Tuple

from scope.engine.discover import is_test_file as _is_test_file
from scope.models import Symbol

_IDENTIFIER_RE = re.compile(r"[a-zA-Z_]\w+")


def build_symbol_index(all_symbols: Dict[str, List[Symbol]]) -> Dict[str, List[Symbol]]:
    """Build a token → symbols lookup for fast reference scanning."""
    index: Dict[str, List[Symbol]] = defaultdict(list)
    for symbols in all_symbols.values():
        for sym in symbols:
            for token in _reference_tokens(sym.name):
                index[token].append(sym)
    return dict(index)


def _reference_tokens(name: str) -> Set[str]:
    base = name.rsplit(".", 1)[-1]
    tokens = {base}
    if "." not in name:
        tokens.add(name)
    return {token for token in tokens if _IDENTIFIER_RE.fullmatch(token)}


def compute_importance(
    all_symbols: Dict[str, List[Symbol]],
    repo_path: str,
    file_inrefs: Optional[Dict[str, int]] = None,
) -> None:
    """Rank symbols by cross-file reference count.

    Reads every source file once, scans for identifier tokens that match known
    symbol base names, and counts how many *other* files reference each symbol.
    """
    index = build_symbol_index(all_symbols)
    token_set = frozenset(index.keys())
    if not token_set:
        return

    ref_count: Dict[Tuple[str, str], int] = defaultdict(int)

    for file_path, symbols in all_symbols.items():
        full_path = os.path.join(repo_path, file_path)
        try:
            with open(full_path, "r", encoding="utf-8", errors="ignore") as f:
                content = f.read()
        except (OSError, IOError):
            continue

        words_in_file: Set[str] = set()
        for word in _IDENTIFIER_RE.finditer(content):
            token = word.group(0)
            if token in token_set and token not in words_in_file:
                words_in_file.add(token)
                for sym in index[token]:
                    if sym.file != file_path:
                        ref_count[(sym.file, sym.name)] += 1

    for file_path, symbols in all_symbols.items():
        is_test = _is_test_file(file_path)

        for sym in symbols:
            score = float(ref_count.get((sym.file, sym.name), 0))
            sym.ref_count = int(score)

            # Boost important symbol kinds
            if sym.kind in ("class", "interface"):
                score *= 1.5
            elif sym.kind in ("resource", "module", "data"):
                score *= 2.0
            elif sym.kind == "key":
                score = -1.0

            # Boost well-known entry point names
            base_name = sym.name.rsplit(".", 1)[-1]
            if base_name in ("main", "index", "App", "Server",
                             "setup", "configure", "create_app", "handler"):
                score += 5.0

            # Boost files with high incoming import counts
            if file_inrefs:
                score += min(file_inrefs.get(file_path, 0), 25) * 0.35

            # Penalize test files
            if is_test or base_name.startswith("test_") or base_name.startswith("it("):
                score *= 0.05

            sym.importance = score


def suggested_reads(all_symbols: Dict[str, List[Symbol]], files: List[str], limit: int = 5) -> List[str]:
    """Rank files by total accumulated symbol importance."""
    scored: List[Tuple[float, str]] = []
    for file_path, symbols in all_symbols.items():
        total = sum(sym.importance for sym in symbols)
        scored.append((total, file_path))
    if not scored:
        return files[:limit]
    scored.sort(key=lambda item: (-item[0], item[1]))
    return [file_path for _score, file_path in scored[:limit]]
